src/domain/core/src/services/pull_request_planner.rs

```diff
@@ -0,0 +1,427 @@
+// Copyright Kamu Data, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use std::borrow::Cow;
+use std::sync::Arc;
+
+use ::serde::{Deserialize, Serialize};
+use internal_error::InternalError;
+use opendatafabric::*;
+use thiserror::Error;
+
+use crate::*;
+
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+// Service
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+#[async_trait::async_trait]
+pub trait PullRequestPlanner: Send + Sync {
+    async fn build_pull_plan(
+        &self,
+        request: PullRequest,
+        options: &PullOptions,
+        tenancy_config: TenancyConfig,
+    ) -> Result<PullPlanIterationJob, PullResponse>;
```

Comment:
Hm, how does this method act when `options.recursive` is `true`? Shouldn't it still return a Vec of iterations?

Reply:
Not allowed. This entry is supposed to solely simplify life of the task executor.

```
#[async_trait::async_trait]
impl PullRequestPlanner for PullRequestPlannerImpl {
    #[tracing::instrument(level = "debug", skip_all, fields(?request, ?options))]
    async fn build_pull_plan(
        &self,
        request: PullRequest,
        options: &PullOptions,
        tenancy_config: TenancyConfig,
    ) -> Result<PullPlanIterationJob, PullResponse> {
        assert!(!options.recursive);
```